integration-tests/test/ovm-self-upgrades.spec.ts

```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  const checkChugsplashInstructionsApplied = async (
+    instructions: ChugSplashInstructions
+  ) => {
+    for (const instruction of instructions) {
+      // TODO: promise.all this for with a map for efficiency
+      if (isSetStorageInstruction(instruction)) {
+        const actualStorage = await l2Provider.getStorageAt(
+          instruction.target,
+          instruction.key
+        )
+        expect(actualStorage).to.deep.eq(instruction.value)
+      } else {
+        const actualCode = await l2Provider.getCode(instruction.target)
+        expect(actualCode).to.deep.eq(instruction.code)
+      }
+    }
+  }
+
+  const applyAndVerifyUpgrade = async (
+    instructions: ChugSplashInstructions
+  ) => {
+    await applyChugsplashInstructions(instructions)
+    await checkChugsplashInstructionsApplied(instructions)
+  }
+
+  before(async () => {
+    env = await OptimismEnv.new()
+    l2Wallet = env.l2Wallet
+
+    OVM_UpgradeExecutor = new Contract(
+      '0x420000000000000000000000000000000000000a',
+      getContractInterface('OVM_UpgradeExecutor', true),
+      l2Wallet
+    )
+
+    Factory__ReturnOne = await ethers.getContractFactory('ReturnOne', l2Wallet)
+    const Factory__ReturnTwo = await ethers.getContractFactory(
+      'ReturnTwo',
+      l2Wallet
+    )
+    const returnTwo = await (
+      await Factory__ReturnTwo.deploy()
+    ).deployTransaction.wait()
+    DeployedBytecode__ReturnTwo = await l2Provider.getCode(
+      returnTwo.contractAddress
+    )
+
+    Factory__SimpleStorage = await ethers.getContractFactory(
+      'SimpleStorage',
+      l2Wallet
+    )
+  })
+
+  describe('setStorage and setCode are correctly applied according to geth RPC', () => {
+    it('Should execute a basic storage upgrade', async () => {
+      const basicStorageUpgrade: ChugSplashInstructions = [
+        {
+          target: OVM_ETH_ADDRESS,
+          key:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+          value:
+            '0x6789123412341234123412341234123412341234123412341234678967896789',
+        },
+      ]
+      await applyAndVerifyUpgrade(basicStorageUpgrade)
+    })
+
+    it('Should execute a basic upgrade overwriting existing deployed code', async () => {
+      const DummyContract = await (
+        await ethers.getContractFactory('SimpleStorage', l2Wallet)
+      ).deploy()
+      await DummyContract.deployTransaction.wait()
+
+      const basicCodeUpgrade: ChugSplashInstructions = [
+        {
+          target: DummyContract.address,
+          code:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+        },
+      ]
+      await applyAndVerifyUpgrade(basicCodeUpgrade)
+    })
+
+    it('Should execute a basic code upgrade which is not overwriting an existing account', async () => {
+      // TODO: fix me?  Currently breaks due to nil pointer dereference; triggerd by evm.StateDB.SetCode(...) in ovm_state_manager.go ?
```

Comment:
Could you give an update on this comment?

Reply:
Removing the comment--I have not been able to repro since the first couple days I was implementing this.  Extensive testing is definitely needed :)